### DISCUSSION TITLE
[Modding] Where to get started with modding?

### DISCUSSION THREAD

**Discussion by Conmiester** [diasurgical/DevilutionX#6482] [Q&A] [answered]
Hello! Since I had a blast with DevilutionX for Diablo I + Co-op from a friend of mine recently, I thought about making a mod by using this port. At first, it was going to be small and restore the cut content such as quests but just as I explored several other Diablo 1 mods + few that are related to DevilutionX, it turned into a big one with another set of features that may as well be considered another huge mod for Diablo. However, just for the sake of me being very new with Diablo’s modding scene, I have questions that are more relevant in terms of the scope of my mod for the time being.

1. Is it possible to re-create all of the cut quests that Belzebub has already done but in DevilutionX port?

2. Is it possible to create new monsters and restore the scrapped ones so they can be used in every existing level + custom ones as well?

3. For new other areas that aren’t part of the cut quests, can something like an entrance section of the cut Hall of the Heroes be implemented in either first 4 levels of the Cathedral?

4. Can there be a way to implement fully shared EXP and maybe an option that disables dropping gear and gold upon death in Multiplayer? (the second one is a personal choice so I may leave it as it is or change it regardless)

5. For Diablo, is it possible to make him become a unique monster so he can drop unique items and make a new item as a soul stone upon right clicking, the ending triggers instead of the game ending by itself after killing him?

6. Last and not least, Is it possible to create other stuff? Such as…

- Being able to run in dungeons but it gets disabled when your in combat against an enemy
- Alternate weapon slot and being able to switch between weapons
- Creating new items and maybe restore some of the scrapped spells, similarly what Belzebub did
- Being able to modify the balance gameplay like what Infernity and Apocrypha does

There are other things I could have asked that may as well be crazy ideas that could be created in my mod such as custom classes like a mage that summons minions, assassin-ish class that uses invisibility as her skill and a caster that uses blood magic and only uses vitality stat and 4 more skills for each class similarly like Belzebub’s but those will have to wait. That being said, is it all possible in DevilutionX especially with modding tools nowadays?

That is all
Conmiester

**Comment by DakkJaniels** [comment]
Everything/anything can be done - the project source is available. It just depends on how much effort you want to put into making changes.

**Reply by Conmiester** [reply]
Honestly, it varies on how much I can put effort into it if I've understood enough of how it works but at this moment who I have no knowledge with how Diablo 1 modding works and haven't found out how to setup the entire mod structure and then get into it especially from GitHub perspective, so I pretty much hit a roadblock ATM. 

Then again, I'm not really a fan of being a lone modder and lacking outside help from others which admittedly, it has lasted for a long time. _It's a pain._

**Reply by DakkJaniels** [reply]
Any substantial modding requires editing the C++ source code and building a new binary. You could do graphics/color swaps without editing the source, that's probably about it.

**Reply by AJenbo** [reply]
Some light changing of levels is also doable and soon more.

**Comment by AJenbo** [comment, marked as answer]
Graphics tool: https://github.com/diasurgical/d1-graphics-tool
Level editor: https://github.com/diasurgical/modding-tools/tree/master/Tiled
Shared xp: https://github.com/diasurgical/devilutionX/pull/6293
5: yes, but also you can have him spawn what every you like regardless, requires coding.
6:
- I think Infernity did this
- Requres coding
- Hopefully will not require coding soon https://github.com/diasurgical/devilutionX/pull/6399
- yes... Infernity and Apocrypha did so after all...

If you can write the C++ for it there is no limites to what can be done

**Reply by Conmiester** [reply]
Those two tools will certainly do, considering it's accessibility looks good to my liking. Hopefully it'll be enough to get the hang of them, apart from a few things to consider for the cut quests levels especially the Horazon's Sanctum (aka, my most favorite restored cut quest from the series by Belzebub). Perhaps time will tell someday.

Shared XP: So it seems shared XP is possible which it's good! I could be wrong but it seems that it could be used as a asset for modders which that's fine if that's the case.
5: That's alright but given my lack or no experience of with Diablo 1's code, it may not take much to edit. Though, what I meant about Diablo's soulstone as an item, it's this link that showcases it: [https://youtu.be/renjD2-8isM](url)
6:

- Kinda funny that I haven't checked out that mod than The Hell 2 mod that had that mechanic which regardless, it's pretty nice.
- Figures.
- Sounds promising, though, seems like that will have to wait till it's official.
- Close enough.

Yeah, but I have no experience with C++ which that's a problem when it comes to coding certain custom mechanics from there so I'm definitely going have to find some help.

**Reply by AJenbo** [reply]
try the chat

**Comment by kphoenix137** [comment]
> Hello! Since I had a blast with DevilutionX for Diablo I + Co-op from a friend of mine recently, I thought about making a mod by using this port. At first, it was going to be small and restore the cut content such as quests but just as I explored several other Diablo 1 mods + few that are related to DevilutionX, it turned into a big one with another set of features that may as well be considered another huge mod for Diablo. However, just for the sake of me being very new with Diablo’s modding scene, I have questions that are more relevant in terms of the scope of my mod for the time being.
> 
>     1. Is it possible to re-create all of the cut quests that Belzebub has already done but in DevilutionX port?

Yes, it absolutely is. I'm involved in a project that restores all the cut quests. I found that it's quite difficult and time consuming. So far I've just restored the Butcher's Chamber quest. This involves using both Tiled and changing/adding code. Creating Horazon's Sanctum would be much more involved as you'd need to add brand new graphics like Belzebub did, as well as figuring out how to create the moving star background.

> 
>     2. Is it possible to create new monsters and restore the scrapped ones so they can be used in every existing level + custom ones as well?

Yes. Very possible. Just requires more code. You'll need understanding of how "Monster size" works though.

> 
>     3. For new other areas that aren’t part of the cut quests, can something like an entrance section of the cut Hall of the Heroes be implemented in either first 4 levels of the Cathedral?
> 
>     4. Can there be a way to implement fully shared EXP and maybe an option that disables dropping gear and gold upon death in Multiplayer? (the second one is a personal choice so I may leave it as it is or change it regardless)

Yes. Also I created a PR awhile back that implements the Diablo 2 style of sharing exp, but it's incomplete as it's not based on range yet. Preventing gear/gold drop is entirely possible, and a hell of a lot easier than point No. 1.

> 
>     5. For Diablo, is it possible to make him become a unique monster so he can drop unique items and make a new item as a soul stone upon right clicking, the ending triggers instead of the game ending by itself after killing him?

Yes, I've done PRs for both of these things. Relevant code can found in them, if you find them. They're likely both closed/unmerged.

> 
>     6. Last and not least, Is it possible to create other stuff? Such as…
> 
> 
>     * Being able to run in dungeons but it gets disabled when your in combat against an enemy

Yes.

> 
>     * Alternate weapon slot and being able to switch between weapons

Yes

> 
>     * Creating new items and maybe restore some of the scrapped spells, similarly what Belzebub did

Yes, I've also personally restored Doom Serpents.

> 
>     * Being able to modify the balance gameplay like what Infernity and Apocrypha does

Yes


> 
> 
> There are other things I could have asked that may as well be crazy ideas that could be created in my mod such as custom classes like a mage that summons minions, assassin-ish class that uses invisibility as her skill and a caster that uses blood magic and only uses vitality stat and 4 more skills for each class similarly like Belzebub’s but those will have to wait. That being said, is it all possible in DevilutionX especially with modding tools nowadays?
> 
> That is all Conmiester

Like others here said, just about _anything_ is possible. However it's all going to take knowledge of C++; that is a given. You cannot do any modding without C++ unless it's strictly doing something like swapping out certain graphics for other graphics, however even that might have cases where code is required.

**Reply by AJenbo** [reply]
You can also swap sounds and some map parts with out code.
